### ISSUE TITLE
Identify and separate the setup_keymanager into deferred calls

### ISSUE THREAD

**Issue by re-nobre** [pixelated/pixelated-user-agent#763]
## What
While working on the issue #759, @bwagnerr split the _get_leap_session call into two smaller ones, we now have setup_keymanager and _create_leap_session.

setup_keymanager is still taking more than our threshold of 100ms and we have to break it apart further. We run on @thaissiqueira machine using vagrant and we got 1.35786 seconds for this call. 

## Definition of done
The goal for this issue is:

- Try to split the setup_keymanager into deferred calls.
- If these new deferred calls still take more than 100ms. Try to split it again.
- If it's not possible to split, separate them out from the main processing thread, like the following:
  - if it's a processing intense call, than we should defer to a process using spawnProcess
  - if it's a disk IO, than we should defer to a new thread using deferToThread
  - if it's a HTTP request, than we should use twisted.web.client

**Comment by olabini** [comment]
I disagree with this definition - the goal here is NOT to decrease the time. The goal is to decrease the time this function and its parts blocks the main reactor thread. This is NOT a performance goal, it is a scalability goal.

**Comment by mnandri** [comment]
So given that Bruno has already moved this part to a  deferToThread, we proceed to check if there isn't any CPU bound tasks inside.

Our measurement  (single user - on dev):
```
Unknown 0.00003 nicknum url
Unknown 0.00015 open pgp
Unknown 0.00022 async client pinned
Unknown 0.00067 async client
Unknown 0.02331 create combined bundle
Unknown 0.02470 keymanager instantiation
Unknown 0.03338 keymanager
Unknown 0.25213 soledad
```
which reveals that the keymanager instantiation is the biggest culprit. Drilling down further, it shows that the  `create combine bundle`  is the responsible.

Reading the [code shows](https://github.com/pixelated/keymanager/blob/develop/src/leap/keymanager/__init__.py#L157-L173) that step: 

```
        with open(tmp_file.name, 'w') as fout:
            fin = fileinput.input(files=(leap_ca_bundle, self._ca_cert_path))
            for line in fin:
                fout.write(line)
            fin.close()
```

is really reading two files, and merging them into one... so mainly File IO.
Drilling further, we found out that these are the very same file for every user.
```
venv/src/leap.common/src/leap/common/cacert.pem
~/.leap/providers/dev.pixelated-project.org/keys/client/api.pem
```
The next step will be to move that operation out of the constructor  and doing it only once during the server boostrap.

**Comment by mnandri** [comment]
We investigated further as the two numbers: 1. from direct measurement and 2. from twisted patch did not add up... so we lowered the threshold in the twisted patch and we obtain the following:
```
0.01034 EXECUTING <generator o2016-08-23 18:00:36 [twisted] INFO processing doclist
0.01149 EXECUTING <generator object get_sync_info at 0x7fe658779b90> in thread 140627744806656
0.01245 EXECUTING <bound method DeferredList._cbDeferred of <DeferredList at 0x7fe6587e1ef0>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:883) in thread 140627744806656
0.01560 EXECUTING <bound method _ThreePhaseEvent._continueFiring of <twisted.internet.base._ThreePhaseEvent object at 0x7fe66ca134d0>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/base.py:420) in thread 140627744806656
0.01929 EXECUTING <bound method OpenPGPScheme._migrate_documents_schema of <leap.keymanager.openpgp.OpenPGPScheme object at 0x7fe6587c5310>> (/root/junk/git_env/src/leap.keymanager/src/leap/keymanager/openpgp.py:110) in thread 140627744806656
0.02051 EXECUTING <generator object init_indexes at 0x7fe6680501e0> in thread 140627744806656
0.02093 EXECUTING <function execute at 0x7fe658721ed8> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:1349) in thread 140627744806656
0.02183 EXECUTING <function send at 0x7fe6587d0140> (/root/junk/git_env/src/leap.keymanager/src/leap/keymanager/__init__.py:352) in thread 140627744806656
0.02227 EXECUTING <bound method DeferredResource._cbChild of <twisted.web.util.DeferredResource instance at 0x7fe65833f5f0>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/web/util.py:143) in thread 140627744806656
0.02230 EXECUTING <function execute at 0x7fe65872b320> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:1349) in thread 140627744806656
0.02337 EXECUTING <function execute at 0x7fe6587ade60> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:1349) in thread 140627744806656
0.02654 EXECUTING <bound method DeferredResource._ebChild of <twisted.web.util.DeferredResource instance at 0x7fe669960ef0>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/web/util.py:146) in thread 140627744806656
0.03020 EXECUTING <function execute at 0x7fe6587eba28> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:1349) in thread 140627744806656
0.03174 EXECUTING <function execute at 0x7fe6587bba28> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:1349) in thread 140627744806656
0.03257 EXECUTING <generator object create_leap_session at 0x7fe668050500> in thread 140627744806656
0.03268 EXECUTING <generator object create at 0x7fe6680503c0> in thread 140627744806656
0.03856 EXECUTING <bound method DeferredResource._cbChild of <twisted.web.util.DeferredResource instance at 0x7fe6680a7638>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/web/util.py:143) in thread 140627744806656
0.06887 EXECUTING <generator object _setup_search_engine at 0x7fe6586ec5a0> in thread 140627744806656
0.06902 EXECUTING <generator object get_or_create_key at 0x7fe6586ec5f0> in thread 140627744806656
0.06957 EXECUTING <generator object _create_new_session at 0x7fe668050550> in thread 140627744806656
0.06968 EXECUTING <generator object setup_keymanager at 0x7fe668050af0> in thread 140627744806656
0.18369 EXECUTING <bound method _ThreePhaseEvent._continueFiring of <twisted.internet.base._ThreePhaseEvent object at 0x7fe66ca13490>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/base.py:420) in thread 140627744806656
0.32719 EXECUTING <generator object _index_all_mails at 0x7fe668050500> in thread 140627744806656
0.32728 EXECUTING <generator object all_mails at 0x7fe6586ec5f0> in thread 140627744806656
0.32736 EXECUTING <generator object all_mails at 0x7fe6586ec5a0> in thread 140627744806656
0.32744 EXECUTING <generator object get_mails at 0x7fe6586ec8c0> in thread 140627744806656
0.32750 EXECUTING <bound method DeferredList._cbDeferred of <DeferredList at 0x7fe658682170>> (/root/junk/git_env/local/lib/python2.7/site-packages/twisted/internet/defer.py:883) in thread 140627744806656
0.32919 EXECUTING <generator object get_mail at 0x7fe6586ec7d0> in thread 140627744806656
0.32935 EXECUTING <generator object _leap_message_to_leap_mail at 0x7fe6586ecfa0> in thread 140627744806656
0.32943 EXECUTING <generator object _mailbox_name_from_uuid at 0x7fe65868a0f0> in thread 140627744806656
0.32955 EXECUTING <generator object _mailbox_uuid_to_name_map at 0x7fe65868a140> in thread 140627744806656
```
and the corresponding direct measurement is:
```
Unknown 0.00003 nicknum url
Unknown 0.00019 open pgp
Unknown 0.00057 async client pinned
Unknown 0.00146 async client
Unknown 0.04904 create combined bundle
Unknown 0.05187 keymanager instantiation
Unknown 0.05632 keymanager
Unknown 0.27606 soledad
```
which made us realise that the [open pgp line](https://github.com/pixelated/keymanager/blob/develop/src/leap/keymanager/__init__.py#L127) in the key manager instantiation is actually a deferred (normal one, not used with @inlineCallbacks) 
as is shown in the code:
```
        self.deferred_init = init_indexes(soledad)
        self.deferred_init.addCallback(self._migrate_documents_schema)
        self._wait_indexes("get_key", "put_key", "get_all_keys")
```
with this, the numbers add up
```
0.01929 EXECUTING <bound method OpenPGPScheme._migrate_documents_schema of <leap.keymanager.openpgp.OpenPGPScheme object at 0x7fe6587c5310>> (/root/junk/git_env/src/leap.keymanager/src/leap/keymanager/openpgp.py:110) in thread 140627744806656
```
and 
```
Unknown 0.04904 create combined bundle
```
explains the total time:
```
0.06968 EXECUTING <generator object setup_keymanager at 0x7fe668050af0> in thread 140627744806656
```
More importantly,  the lines
```
        self.deferred_init.addCallback(self._migrate_documents_schema)
        self._wait_indexes("get_key", "put_key", "get_all_keys")
```
in the OpenPGPScheme constructor, are callbacks and therefore runs on the main threads.
These will be investigated first as they are blocking the main thread.

**Comment by re-nobre** [comment]
After the discussion on the issue #762 , I'm changing this issue's name and description.

**Comment by mnandri** [comment]
Current status,

After running a number of requests (single user on /login) we have observerd that the call on `OpenPGPScheme._migrate_documents_schema` is always small and not blocking the reactor. In fact, we have observed that the direct measurement on keymanager setup and with the openpgpscheme actually do **not** add up to the one measured by the twisted patch (the one above is an exception)... we repeated the experiment multiple times.
At this point, there is no candidate that could actually block in keymanager setup... We suspect that  time measurement for `generator` from the twisted patch has to be interpreted differently from normal `function` callbacks.
@bwagnerr  and @thaissiqueira  is working on this: understand the meaning of the measurements obtained for `generator` in the twisted patch.

In the meantime, we assumed that the original patch measures the accurate time for `function` callbacks. We proceeded then in making the setup of keymanager to be an actual call back. The original [create new session](https://github.com/pixelated/pixelated-user-agent/blob/master/service/pixelated/config/sessions.py#L41) is split into 3 parts (see [code here](https://github.com/pixelated/pixelated-user-agent/blob/cd04eb1b3726dfe12d2f21c5a8d3aef3f4569a28/service/pixelated/config/sessions.py#L48-L67)):
1. fire deferred soledad instantiation
2. add setup keymanager as a callback
3. add the rest of the cod to be a callback

We obtain the following results:
```
0.01258 EXECUTING <function setup_keym at 0x7f964cbdc758> (/vagrant/service/pixelated/config/sessions.py:50) in thread 140283799688960
1.23237 EXECUTING <function the_rest at 0x7f964cbdc668> (/vagrant/service/pixelated/config/sessions.py:55) in thread 140283799688960
1.25510 EXECUTING <generator object setup_soledad at 0x7f964c361730> in thread 140283799688960
```
Assuming measuring actual call back in the twisted patch is correct, this would mean that the setup of keymanager is 100 times faster than the rest of the code. It is not blocking the reactor.

For the sake of curiosity, we split it further for every line to be on its own callback (see [code here](https://github.com/pixelated/pixelated-user-agent/blob/9b5f232f31ed4a242d141bc5595648812d9e7f46/service/pixelated/config/sessions.py#L48-L84)), we obtain the following:

```
0.01235 EXECUTING <function setup_keym at 0x7f64c4188758> (/vagrant/service/pixelated/config/sessions.py:50) in thread 140070944990976
1.23972 EXECUTING <function _smtp_info at 0x7f64c4188848> (/vagrant/service/pixelated/config/sessions.py:64) in thread 140070944990976
1.26689 EXECUTING <generator object setup_soledad at 0x7f64c418e6e0> in thread 140070944990976
```
We isolated the block to be in the `self._provider.smtp_info()`. This actually has a blocking GET request.

Note: setup_soledad is a generator (from @defer.inlineCallbacks -- with yield), we would need the result of the investigation to be able to state on the meaning of the actual time measured.

**Comment by thaissiqueira** [comment]
Analysing how the twisted patch measures timings, @bwagnerr and I believe that this patch is not suitable to measure the inlineCallbacks functions we found.

We got that suspicion because the timings we got for the setup_keymanager, that only calls a deferToThread, was taking 1.27 seconds. This is the function:

```python
    @defer.inlineCallbacks
    def setup_keymanager(self, provider, soledad, account_email, token, uuid):
        keymanager = yield threads.deferToThread(Keymanager,
                                                 provider,
                                                 soledad,
                                                 account_email,
                                                 token,
                                                 uuid)
        defer.returnValue(keymanager)
```

We analysed the patch to understand the behaviour of the timings when used with inlineCallbacks and came to the belief that our patch can only measure correctly traditional defer.addCallback calls.

When used with inlineCallbacks, the timing of the patch includes everything from the start of the inlineCallback invocation to the defer.returnValue(x) at the end. This includes all the yields that interrupt the method execution and allow other things to run on the reactor.

With this we concluded that the current patch can't be used to time inlineCallbacks, we are adapting the current approach to filter the inlineCallbacks timing and are working on adding a timing for the [inlineCallback g.send(result)] (https://github.com/twisted/twisted/blob/trunk/src/twisted/internet/defer.py#L1169), that resumes the execution of the inlineCallback generator until the next yield interruption or the end of the method.

We'll now revise the findings after the split on the issue #759 and figure out how to go forward

@olabini Does this approach make sense to you?

**Comment by olabini** [comment]
I don't know how inlineCallback is implemented. My assumption was that it used the regular Deferred class for implementation, and I assumed my initial patch would have covered that as well.

Not sure which the current patch is you are referring to.

But yeah, it seems clear that if inlineCallback doesn't use the regular Deferred class, we need something else to measure that. So that sounds reasonable. =)

**Comment by thaissiqueira** [comment]
Measuring inlineCallbacks (with the patch created on #768 issue), we got more accurate measurements of the setup_keymanager call timing on the main thread.
With this we concluded that the setup_keymanager execution on the main thread is under 100ms and we are deprioritizing this issue.

**Comment by tuliocasagrande** [comment]
So we moved the last action to #765, right?
@mnandri

**Comment by deniscostadsc** [comment]
@tuliocasagrande  @mnandri 

Can I unassign you from this issue?

**Comment by thaissiqueira** [comment]
The setup_keymanager is faster than 0.1 seconds. So, it's not a blocker for now.
Because of that I'm closing this issue.
